Refactor download embedded in provider class

tgaugry opened the pull request:
Hello,
following the questions on #360, here come the PR !

As the changes are pretty consequents, more so than I imagined, this is for now a "draft" PR.

Changes:
- `Provider` class is stripped out of the actual download. It is defered to the classes inheriting `BaseDownloadMethod`. This simplify greatly the following changes.
- `process` method is now split into to 2 methods `prepare_download` and `loop_chapters`
- `Provider` now use by default the `OneByOneDownloadMethod`, which download file by file. Most (if not all) of the code comes from what was inside `Provider` class.
- `Http2` code has been move to  `WholeArchiveDownloadMethod` and `Process`
- As a result, the classes using Http2 now use the `WholeArchiveDownloadMethod`. They no longer have to override `loop_chapter`, and now use a consistently named `download_chapter` method.
- The classes using `Http2` were using a different kind of progress bar. The functionality was kept, but moved to the `--global-progress` command line flag instead.
- The `Http2` tests were made on manga_online_biz website. The others either didn't exist, or seem to require an account.
- manga_online_biz  was not working initially. Now it does. (ex: `manga-py -s 3 -c2 --show-chapter-info https://manga-online.biz/martial_peak_12.html`)

Remaining work/questions/etc:
- The others `Http2` were not tested
- The download methods classes can likely be further refactored.
- Archives Naming of the classes that used `Http2` has changed. I don't know why it was different previously though. For example, manga_online_biz  return 2 numbers as chapter index, and i can't fathom why.
- The new classes use `self.provider` an awful lot.
- Maybe this is a good time to split the one by one method into a sequential and a parallel (current) version, as some providers don't behave nicely with a lot of requests (looking at you mangadex !)
- Naming could be improved, I'm horrible at that ^^

tgaugry committed:
provider: Split process method into prepare_download and loop_chapters

tgaugry committed:
Move _min_max_calculate to provider

tgaugry committed:
- Simplify _min_max_calculate
- Remove unnecessary parameters

tgaugry committed:
Add global progress/chapter progress

tgaugry committed:
Move the downloading out of provider - Introduce the download_methods classes

1271 commented:
Thank you! I'll try to fix the providers and merge it then

1271 commented:
> manga_online_biz return 2 numbers as chapter index

Chapter and part. Example: ch_1_vol_2.zip

> Maybe this is a good time to split the one by one method into a sequential and a parallel

Initially, there were 2 methods of loading, then I refused. You can make manga-py load 1 image by passing `--max-threads 1`

tgaugry commented:
> Thank you! I'll try to fix the providers and merge it then

If you have a command line that should be working with these providers, I can test that while doing the remaining changes/cleanup.
Don't merge yet though. I still have some changes to do (mostly cleanup, can't wrap my head around these `self.provider` i've added everywhere). I'll remove the draft flag when it's ready (this weekend, likely). Just wanted to show the bulk/structure of the changes.

> > manga_online_biz return 2 numbers as chapter index
> 
> Chapter and part. Example: ch_1_vol_2.zip

Hoooo I see, so it goes along the `--zero-fill` option.  I didn't test that, will do/fix if needed.

> > Maybe this is a good time to split the one by one method into a sequential and a parallel
> 
> Initially, there were 2 methods of loading, then I refused. You can make manga-py load 1 image by passing `--max-threads 1`

Seems fair. I'll refrain from doing that change then.

1271 commented:
No problems. Thanks for your work 👍 

> Hoooo I see, so it goes along the --zero-fill option. I didn't test that, will do/fix if needed

Yeah

tgaugry committed:
Remove duplicate before/after_download

tgaugry committed:
Simplify Callback - remove _call_files_progress_callback; instead set a default for all if callable is not valid
Default number of threads (4) is now set in the cli parser
Parallel download now use concurent.future.ThreadPool instead
Progress callback are now done in threadpool.map call instead of inside the saving functions

tgaugry commented:
So `--zero-fill` is working as intended.

 However, some provider override `save_file` and I missed that.
The list:
- e_hentai_org -> Works
- japscan_com -> "Provider not worked"
- allhentai_ru -> No new chapters found; also looks like chapter can be dowloaded as archive ?
- plus_comico_jp -> Cannot test (no free manga ?)
- plus_comico_jp_manga -> Cannot test (no free manga ?)
- viz_com -> Cannot test (no free manga ?)
- cdmnet_com_br -> Not working, but fix seems easy
- shogakukan_tameshiyo_me -> already deactivated in `providers/__init__.py`

They currently break with this PR.
So still not ready yet !

EDIT: Ok, most of them are already broken without the PR. I'll see if they can be fixed.

1271 commented:
There are free chapters for comico_jp / viz_com, but only for the USA (I had to rent a vpn to check these providers) xD

allhentai_ru - Yes, there are several providers that download the archive instead of images.

You can find them by `return [b'']`

tgaugry committed:
Fix _min_max

tgaugry committed:
Fix e_hentai_org
Fix allhentai_ru chapter list - still not working, website has changed

tgaugry committed:
Fix cdmnet_com_br link
Add japscan new extension

tgaugry committed:
Add new links

tgaugry commented:
This should fix most if not all of them.
The `return [b'']` providers are  not always archive dowload, some of them are just one shots, but still downloaded image per image. These are not broken by my PR. The are almost all porn btw, this will be really nice on my github profile :D. 

In summary:

Provider | Current (1.25.0) | PR
---------|------------------|----
e_hentai_org  | Works | Works
japscan_com | Broken |
allhentai_ru | Broken | Less broken
plus_comico_jp | Untested | 
plus_comico_jp_manga | Untested | 
viz_com | Untested | 
cdmnet_com_br | Broken |  Fixed - link changed
shogakukan_tameshiyo_me | Deactivated | 
hentaiporns_net.py | Works | Works
pururin_io  | Broken | 
doujins_com | Broken | 
freeadultcomix_com | Cannot test | 
porncomix_info | Adress changed |
chochox_com | Broken |
hentai_image_com | Wrong link | Fixed
tsumino_com | Wrong link |
hentaifox_com |  Broken |


So this PR don't break any provider that wasn't already broken.
On my side, it seems like this is almost ready, only thing left being the cleanup of the `self.provider`, which should be non breaking.

tgaugry committed:
Cleanup + Remove unwanted log

tgaugry commented:
Ok, i've wracked my brain trying to find a good way to have less calls to `provider`, the only solution i've found would be to have a `chapter` object, which would be separate from the provider. In that case, the providers would only generate the `chapter` objects, and any download/archive would be done there.
The problem is that it will require modifying all the providers subclasses, and i'm not too keen on that. So I suggest we leave this change for a later date (if ever at all !).

In any case, i believe it is now ready !

1271 commented:
Ok thanks. I'll watch it this weekend

1271 commented:
OK. Let's merge. If there are any mistakes, I will correct it.
Thanks to everyone who creates the issue. It's better than being silent

tgaugry commented:
Thanks. I'll watch the new issues to see if it has side effects. Cheers.
